### PR TITLE
Convs in Unet on device. Support for 1x1 conv. Fixes.

### PR DESCRIPTION
**Pull request by tt-nshanker**
CI: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7997467579
Running failed tests again after fix: https://github.com/tenstorrent-metal/tt-metal/actions/runs/8000487426
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8000494864
Fixed remaining CI failures locally.

### CODE REVIEW

File: tt_eager/tt_lib/fallback_ops/fallback_ops.py [hunk 1]
```diff
@@ -475,6 +518,7 @@ def __init__(
 
     @convert_tt_tensors_wrapper
     def forward(self, input: ttl_tensor.Tensor) -> ttl_tensor.Tensor:
+        assert False
```
Comment (tt-nshanker) [inline tt_eager/tt_lib/fallback_ops/fallback_ops.py:521]: Added this for local testing to ensure no convs are run on the fallback path. I will remove this later before running CI tests.

File: ttnn/ttnn/__init__.py [hunk 2]
```diff
@@ -112,6 +110,7 @@ def get_bool_env_var(name, default):
     # reduction operations
     mean,
     upsample,
+    unpad,
```
Comment (tt-nshanker) [inline ttnn/ttnn/__init__.py:113]: TODO: Need to move this unpad op to __get_item__

File: models/experimental/functional_stable_diffusion/tt/ttnn_functional_basic_transformer_block.py [hunk 3]
```diff
@@ -88,8 +88,8 @@ def basic_transformer_block(
     )
     if use_ada_layer_norm_zero:
         assert False, "AdaLayerNormZero not supported and not used in stable diffusion"
-
-    ff_output = feedforward(config=config, hidden_states=norm_hidden_states, parameters=parameters.ff)
+    # ttnn.dump_device_memory_state(device)
```
Comment (boris-drazic) [inline models/experimental/functional_stable_diffusion/tt/ttnn_functional_basic_transformer_block.py:91]: Can you delete comment?

File: models/experimental/functional_stable_diffusion/tt/ttnn_functional_downsample_2d.py [hunk 4]
```diff
@@ -5,9 +5,28 @@
...
+    input = ttnn.to_layout(input, layout)
+    input = ttnn.to_device(input, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
+    return input
+
+
+def ttnn_to_torch(input):
+    input = ttnn.to_layout(input, ttnn.ROW_MAJOR_LAYOUT)
+    input = ttnn.from_device(input)
+    input = ttnn.to_torch(input)
+    return input
```
Comment (boris-drazic) [inline models/experimental/functional_stable_diffusion/tt/ttnn_functional_downsample_2d.py:29]: Can you put these two functions in utility_functions file?

File: models/experimental/functional_stable_diffusion/tt/ttnn_functional_transformer_2d.py [hunk 5]
```diff
@@ -4,12 +4,31 @@
...
+    input = ttnn.to_layout(input, layout)
+    input = ttnn.to_device(input, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
+    return input
+
+
+def ttnn_to_torch(input):
+    input = ttnn.to_layout(input, ttnn.ROW_MAJOR_LAYOUT)
+    input = ttnn.from_device(input)
+    input = ttnn.to_torch(input)
+    return input
```
Comment (boris-drazic) [inline models/experimental/functional_stable_diffusion/tt/ttnn_functional_transformer_2d.py:31]: Can you put these two functions in utility_functions file?

File: models/experimental/functional_stable_diffusion/tt/ttnn_functional_upsample_nearest_2d.py [hunk 6]
```diff
@@ -6,9 +6,22 @@
...
+    # up_output = ttnn.repeat_interleave(up_output, scale_factor, dim=2)
+
+    print(f"=============================== input shape: {input.shape}")
+
+    ## permute to NHWC
+    input = ttnn.to_layout(input, ttnn.ROW_MAJOR_LAYOUT)
+    input = ttnn.permute(input, (0, 2, 3, 1))
+
+    print(f"=============================== input shape: {input.shape}")
+    up_output = ttnn.upsample(input, scale_factor)
```
Comment (boris-drazic) [inline models/experimental/functional_stable_diffusion/tt/ttnn_functional_upsample_nearest_2d.py:22]: Can you remove printf?

**Review by boris-drazic — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/test_resnet50_untilize_with_halo_and_conv_v2.py [hunk 7]
```diff
@@ -654,6 +654,7 @@ def test_resnet50_conv(
             output_dtype=activations_dtype,
             math_fidelity=math_fidelity,
             use_shallow_conv_variant=(C == 16),
+            deallocate_input=True,
```
Comment (tt-nshanker) [inline tests/tt_eager/python_api_testing/unit_testing/test_resnet50_untilize_with_halo_and_conv_v2.py:657]: TODO: fix the variable name

File: ttnn/cpp/ttnn/operations/binary.hpp [hunk 8]
```diff
@@ -67,6 +68,9 @@ inline ttnn::Tensor add(const ttnn::Tensor& input_tensor_a_arg,
             auto input_tensor_b_4D = ttnn::unsqueeze_to_4D(input_tensor_b);
 
             if (height_b == 1 or width_b == 1) {
+                if (dtype.has_value()) {
+                    TT_THROW("Bcast: Unsupported dtype argument.");
```
Comment (arakhmati) [inline ttnn/cpp/ttnn/operations/binary.hpp:72]: "ttnn.add: cannot change dtype when broadcasting");"

File: ttnn/ttnn/operations/core.py [hunk 9]
```diff
@@ -788,6 +790,39 @@ def impl(tensor):
...
+    clone(tensor: ttnn.Tensor, memory_config: MemoryConfig, dtype: DataType) -> ttnn.Tensor
+    Clones the tensor by copying it with the given `memory config`. Also, converts the dataype to `dtype`.
+    Note: clone does not change the layout of the tensor.
+    Organizes the `ttnn.Tensor` :attr:`tensor` into either ROW_MAJOR_LAYOUT or TILE_LAYOUT.  When requesting ROW_MAJOR_LAYOUT
+    the tensor will be returned unpadded in the last two dimensions.   When requesting TILE_LAYOUT the tensor will be automatically
+    padded where the width and height become multiples of 32.
+    In the case where the layout is the same, the operation simply pad or unpad the last two dimensions depending on layout requested.
+    Args:
+        * :attr:`tensor`: the ttnn.Tensor
+        * :attr:`layout`: the layout of either ttnn.ROW_MAJOR_LAYOUT or ttnn.TILE_LAYOUT.
```
Comment (arakhmati) [inline ttnn/ttnn/operations/core.py:817]: add an entry for dtype

**Review by arakhmati — CHANGES_REQUESTED** [review]

**Review by arakhmati — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]

File: tests/ttnn/integration_tests/stable_diffusion/test_unet_2d_condition_model.py [hunk 10]
```diff
@@ -165,4 +165,4 @@ def test_unet_2d_condition_model_512x512(device, batch_size, in_channels, input_
         config=config,
     )
     ttnn_output = ttnn_to_torch(ttnn_output)
-    assert_with_pcc(torch_output, ttnn_output, pcc=0.99)
+    assert_with_pcc(torch_output, ttnn_output, pcc=0.80)
```
Comment (mywoodstock) [inline tests/ttnn/integration_tests/stable_diffusion/test_unet_2d_condition_model.py:169]: I had reduced this, can you revert?

File: ttnn/ttnn/operations/binary.py [hunk 11]
```diff
@@ -164,8 +164,19 @@ def add(
...
+        else:
+            raise TypeError("ttnn.add: cannot change dtype when broadcasting")
     input_tensor_b = input_tensor_b.value if isinstance(input_tensor_b, ttnn.Tensor) else input_tensor_b
-    output = ttnn._ttnn.operations.binary.add(input_tensor_a, input_tensor_b, memory_config=memory_config, dtype=dtype)
+    if run_with_dtype:
+        output = ttnn._ttnn.operations.binary.add(
+            input_tensor_a, input_tensor_b, memory_config=memory_config, dtype=dtype
+        )
+    else:
+        output = ttnn._ttnn.operations.binary.add(input_tensor_a, input_tensor_b, memory_config=memory_config)
```
Comment (arakhmati) [inline ttnn/ttnn/operations/binary.py:179]: Update the C++ function that is overloaded for scalar to take in `dtype` and add the assert there

**Review by arakhmati — CHANGES_REQUESTED** [review]

**Review by arakhmati — APPROVED** [review]

**Pull request merged**